### PR TITLE
Allow code pre-generation to accept files outside of chip-root

### PR DESCRIPTION
**Pull request by andy31415**
This allows passing `external-root` to the script to include codegen outside $CHIP_ROOT.

Updated documentation to describe how to generate `.matter` files and how to use this new option.

### CODE REVIEW

**Review by yunhanw-google — APPROVED** [review]

File: docs/code_generation.md [hunk 1]
```diff
@@ -234,9 +234,48 @@ scripts/codepregen.py ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
...
-scripts/codepregen.py --input-glob "*all-clusters*" ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
+scripts/codepregen.py --input-glob "*all-clusters*" --input-glob "*controller*" ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
 ```
 
+### External applications/zap files
+
+#### Ensure you have a `.matter` file
+
+Code generation generally will use both `.zap` or `.matter` files. If you only
+have a `.zap` file, you can create the corresponding matter via:
```
Comment (bzbarsky-apple) [inline docs/code_generation.md:245]:
```suggestion
have a `.zap` file, you can create the corresponding `.matter` file via:
```

File: docs/code_generation.md [hunk 2]
```diff
@@ -234,9 +234,48 @@ scripts/codepregen.py ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
...
+
+```bash
+scripts/tools/zap/generate.py ${ZAP_FILE_PATH}
+```
+
+The above will use the template `src/app/zap-templates/matter-idl.json` to
+generate a `.matter`.
+
+Matter files are designed to be human readable. It is recommended to take a look
```
Comment (bzbarsky-apple) [inline docs/code_generation.md:254]:
```suggestion
`.matter` files are designed to be human readable. It is recommended to take a look
```

File: docs/code_generation.md [hunk 3]
```diff
@@ -234,9 +234,48 @@ scripts/codepregen.py ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
...
+at the generated file and see if it contains what is expected and also lint it.
+If anything seems wrong, the `.zap` file should be fixed
+(` .matter`` represents the content of `.zap`). To lint use:
```
Comment (bzbarsky-apple) [inline docs/code_generation.md:257]:
```suggestion
(` .matter` represents the content of `.zap`). To lint use:
```

File: docs/code_generation.md [hunk 4]
```diff
@@ -234,9 +234,48 @@ scripts/codepregen.py ${OUTPUT_DIRECTORY:-./zzz_pregenerated/}
...
+
+```bash
+scripts/idl_lint.py ${MATTER_FILE_PATH}
+```
+
+#### Running pre-generation
+
+If you have zap files outside the CHIP repository (i.e. not in `src` or
+`examples`) you should provide the root of your external root(s) to the script:
```
Comment (bzbarsky-apple) [inline docs/code_generation.md:266]: I don't understand what "root of your external root" means...

Comment (andy31415) [reply docs/code_generation.md]: rephrased.

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**